[TestKit + JUnit]The Gradle daemon holds file descriptors on the binary test-result files

### Current Behavior

On: Windows Server 2019 version 10.0 

The Gradle daemon holds file descriptors on the binary test-result files (test-results/hostTest/binary/*.bin) open after the build completes. JUnit's @TempDir cleanup runs before the daemon releases those handles, causing the deletion to fail.

This leads to a failure with an exception and the following cause:
```
Caused by: java.io.IOException: Failed to delete temp directory Z:\BuildAgent\temp\buildTmp\junit-1267999761028841581. The following paths could not be deleted (see suppressed exceptions for details): <root>, 9.5.1, 9.5.1\6iYNSSXVogOUmLq, 9.5.1\6iYNSSXVogOUmLq\empty, 9.5.1\6iYNSSXVogOUmLq\empty\build, 9.5.1\6iYNSSXVogOUmLq\empty\build\test-results, 9.5.1\6iYNSSXVogOUmLq\empty\build\test-results\jsBrowserTest, 9.5.1\6iYNSSXVogOUmLq\empty\build\test-results\jsBrowserTest\binary
```

### Expected Behavior

@TempDir should be cleaned up, test should pass 

### Context (optional)

There are several connected issues in tests for Kotlin project:
- https://youtrack.jetbrains.com/issue/KT-85872/NativeKGP-testNativeStderrOutputStaysInTestLog-fails-on-Windows-CI-Gradle-daemon-holds-file-descriptors-on-test-results-hostTest
- [KT-87568](https://youtrack.jetbrains.com/issue/KT-87568) JsBrowserTestsWithPlaywrightIT tests fail on Windows

### Self-contained Reproducer Project

Test in Kotlin project -  https://github.com/JetBrains/kotlin/blob/master/libraries/tools/kotlin-gradle-plugin-integration-tests/src/test/kotlin/org/jetbrains/kotlin/gradle/js/JsBrowserTestsWithPlaywrightIT.kt

### Gradle version

9.5.1

### Build scan URL (optional)

_No response_

### Your Environment (optional)

_No response_

--- a/platforms/software/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/report/generic/TestTreeModelResultsProvider.java
+++ b/platforms/software/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/report/generic/TestTreeModelResultsProvider.java
@@ -66,10 +66,9 @@ public final class TestTreeModelResultsProvider implements TestResultsProvider {
     public static void useResultsFrom(Path resultsDir, Consumer<TestTreeModelResultsProvider> resultsConsumer) {
         SerializableTestResultStore resultsStore = new SerializableTestResultStore(resultsDir);
         Serializer<TestOutputEvent> testOutputEventSerializer = TestEventSerializer.create().build(TestOutputEvent.class);
-        try  {
+        try (TestOutputReader outputReader = resultsStore.createOutputReader(testOutputEventSerializer)) {
             TestTreeModel root = TestTreeModel.loadModelFromStores(Collections.singletonList(resultsStore));
-            TestTreeModelResultsProvider resultsProvider = new TestTreeModelResultsProvider(root, resultsStore.createOutputReader(testOutputEventSerializer));
-            resultsConsumer.accept(resultsProvider);
+            resultsConsumer.accept(new TestTreeModelResultsProvider(root, outputReader));
         } catch (Exception e) {
             throw UncheckedException.throwAsUncheckedException(e);
         }

--- a/platforms/software/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/results/serializable/TestOutputReader.java
+++ b/platforms/software/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/results/serializable/TestOutputReader.java
@@ -49,6 +49,7 @@ public final class TestOutputReader implements Closeable {
      * Channels that are currently open and can be reused. We limit the pool size to avoid excessive resource usage.
      */
     private final BlockingQueue<SeekableByteChannel> channelPool = new LinkedBlockingQueue<>(64);
+    private volatile boolean closed;
 
     TestOutputReader(Path outputEventsFile, Serializer<TestOutputEvent> testOutputEventSerializer) {
         this.outputEventsFile = outputEventsFile;
@@ -174,6 +175,9 @@ public final class TestOutputReader implements Closeable {
     }
 
     private SeekableByteChannel requestChannel() {
+        if (closed) {
+            throw new IllegalStateException("This reader has been closed: " + outputEventsFile);
+        }
         SeekableByteChannel open = channelPool.poll();
         if (open != null && open.isOpen()) {
             return open;
@@ -189,7 +193,7 @@ public final class TestOutputReader implements Closeable {
         // Return the channel to the pool for reuse
         boolean added;
         try {
-            added = channelPool.offer(channel);
+            added = !closed && channelPool.offer(channel);
         } catch (Throwable t) {
             // If we fail to return the channel to the pool, close it to avoid resource leaks
             try {
@@ -200,7 +204,7 @@ public final class TestOutputReader implements Closeable {
             throw t;
         }
         if (!added) {
-            // Pool is full, close the channel
+            // The reader was closed while this channel was in use, or the pool is full
             try {
                 channel.close();
             } catch (IOException e) {
@@ -211,6 +215,8 @@ public final class TestOutputReader implements Closeable {
 
     @Override
     public void close() throws IOException {
+        closed = true;
         CompositeStoppable.stoppable(channelPool).stop();
+        channelPool.clear();
     }
 }

--- a/platforms/software/testing-base/src/test/groovy/org/gradle/api/internal/tasks/testing/report/generic/TestTreeModelResultsProviderTest.groovy
+++ b/platforms/software/testing-base/src/test/groovy/org/gradle/api/internal/tasks/testing/report/generic/TestTreeModelResultsProviderTest.groovy
@@ -1,0 +1,84 @@
+/*
+ * Copyright 2026 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.tasks.testing.report.generic
+
+import org.gradle.api.internal.tasks.testing.DecoratingTestDescriptor
+import org.gradle.api.internal.tasks.testing.DefaultTestDescriptor
+import org.gradle.api.internal.tasks.testing.DefaultTestOutputEvent
+import org.gradle.api.internal.tasks.testing.DefaultTestSuiteDescriptor
+import org.gradle.api.internal.tasks.testing.TestCompleteEvent
+import org.gradle.api.internal.tasks.testing.TestStartEvent
+import org.gradle.api.internal.tasks.testing.results.DefaultTestResult
+import org.gradle.api.internal.tasks.testing.results.serializable.SerializableTestResultStore
+import org.gradle.api.tasks.testing.TestOutputEvent
+import org.gradle.api.tasks.testing.TestResult
+import spock.lang.Specification
+import spock.lang.TempDir
+
+import java.nio.file.Path
+
+class TestTreeModelResultsProviderTest extends Specification {
+
+    @TempDir
+    Path tempDir
+
+    def "closes the output events file once the results are no longer in use"() {
+        given:
+        def storeDir = writeStore()
+
+        when:
+        TestTreeModelResultsProvider captured = null
+        long classId = -1
+        def outputWhileInUse = new StringWriter()
+        TestTreeModelResultsProvider.useResultsFrom(storeDir) { provider ->
+            captured = provider
+            provider.visitClasses { classId = it.id }
+            provider.writeAllOutput(classId, TestOutputEvent.Destination.StdOut, outputWhileInUse)
+        }
+
+        then:
+        outputWhileInUse.toString() == "hello"
+
+        when:
+        captured.writeAllOutput(classId, TestOutputEvent.Destination.StdOut, new StringWriter())
+
+        then:
+        thrown(IllegalStateException)
+    }
+
+    private Path writeStore() {
+        def storeDir = tempDir.resolve("store")
+        def writer = new SerializableTestResultStore(storeDir).openWriter(0)
+        try {
+            def root = new DefaultTestSuiteDescriptor(1, "root")
+            def testA = new DecoratingTestDescriptor(new DefaultTestDescriptor(2, "com.example.MySuite", "testA"), root)
+
+            writer.started(root, new TestStartEvent(100))
+            writer.started(testA, new TestStartEvent(100))
+            writer.output(testA, new DefaultTestOutputEvent(100, TestOutputEvent.Destination.StdOut, "hello"))
+            writer.completed(testA, successResult(100, 200), new TestCompleteEvent(200))
+            writer.completed(root, successResult(100, 200), new TestCompleteEvent(200))
+        } finally {
+            writer.close()
+        }
+        return storeDir
+    }
+
+    private static TestResult successResult(long startTime, long endTime) {
+        return new DefaultTestResult(TestResult.ResultType.SUCCESS, startTime, endTime, 1, 1, 0, [], null)
+    }
+}

